### PR TITLE
aya-bpf/maps: Add `get_ptr` and `get_mut_ptr` methods to Array

### PR DESCRIPTION
**Pull request by vadorovsky**
Signed-off-by: Michal Rostecki <vadorovsky@gmail.com>

### CODE REVIEW

**Comment by vadorovsky** [comment]
I will add an integration test once #401 is merged

File: bpf/aya-bpf/src/maps/array.rs [hunk 1]
```diff
@@ -47,14 +47,27 @@ impl<T> Array<T> {
...
-            );
-            // FIXME: alignment
-            NonNull::new(value as *mut T).map(|p| p.as_ref())
-        }
+        unsafe { self.lookup(index).map(|p| p.as_ref()) }
+    }
+
+    #[inline(always)]
+    pub fn get_ptr(&self, index: u32) -> Option<*const T> {
+        unsafe { self.lookup(index).map(|p| p.as_ptr() as *const T) }
```
Comment (alessandrod) [inline bpf/aya-bpf/src/maps/array.rs:57]: I think the extra cast isn't necessary here?

File: bpf/aya-bpf/src/maps/array.rs [hunk 2]
```diff
@@ -47,14 +47,27 @@ impl<T> Array<T> {
...
+    }
+
+    #[inline(always)]
+    pub fn get_ptr_mut(&self, index: u32) -> Option<*mut T> {
+        unsafe { self.lookup(index).map(|p| p.as_ptr() as *mut T) }
```
Comment (alessandrod) [inline bpf/aya-bpf/src/maps/array.rs:62]: as_mut_ptr()

File: bpf/aya-bpf/src/maps/array.rs [hunk 3]
```diff
@@ -47,14 +47,27 @@ impl<T> Array<T> {
...
     }
 
+    #[inline(always)]
     pub fn get(&self, index: u32) -> Option<&T> {
-        unsafe {
-            let value = bpf_map_lookup_elem(
-                self.def.get() as *mut _,
-                &index as *const _ as *const c_void,
-            );
-            // FIXME: alignment
```
Comment (alessandrod) [inline bpf/aya-bpf/src/maps/array.rs:56]: Also you removed the FIXME without fixing it :P

File: bpf/aya-bpf/src/maps/array.rs (hunk 1 above)
Comment (vadorovsky) [reply bpf/aya-bpf/src/maps/array.rs:57]: It is necessary. To my surprise, `NonNull::as_ptr` returns... `*mut T`

https://doc.rust-lang.org/core/ptr/struct.NonNull.html#method.as_ptr

File: bpf/aya-bpf/src/maps/array.rs (hunk 2 above)
Comment (vadorovsky) [reply bpf/aya-bpf/src/maps/array.rs:62]:
```
error[E0599]: no method named `as_mut_ptr` found for struct `NonNull<T>` in the current scope
  --> bpf/aya-bpf/src/maps/array.rs:62:47
   |
62 |         unsafe { self.lookup(index).map(|p| p.as_mut_ptr()) }
   |                                               ^^^^^^^^^^ help: there is a method with a similar name: `as_mut`
```

`as_mut` is not suitable for us, as it returns a reference

https://doc.rust-lang.org/core/ptr/struct.NonNull.html#method.as_mut

File: bpf/aya-bpf/src/maps/array.rs (hunk 3 above)
Comment (vadorovsky) [reply bpf/aya-bpf/src/maps/array.rs:56]: sorry, for now I will put the comment back

but then I will try to actually fix it

File: bpf/aya-bpf/src/maps/array.rs [hunk 4]
```diff
@@ -47,14 +47,28 @@ impl<T> Array<T> {
...
+        unsafe { self.lookup(index).map(|p| p.as_ptr()) }
+    }
+
+    #[inline(always)]
+    unsafe fn lookup(&self, index: u32) -> Option<NonNull<T>> {
+        let ptr = bpf_map_lookup_elem(
+            self.def.get() as *mut _,
+            &index as *const _ as *const c_void,
+        );
+        // FIXME: alignment
```
Comment (alessandrod) [inline bpf/aya-bpf/src/maps/array.rs:71]: the fixme here can go, the pointer doesn't have to be aligned here

**Comment by ajwerner** [comment]
It'd be nice to have this. Any updates on the progress? Would you appreciate help in any way?

File: bpf/aya-bpf/src/maps/array.rs [hunk 5]
```diff
@@ -47,14 +47,30 @@ impl<T> Array<T> {
...
-            // FIXME: alignment
-            NonNull::new(value as *mut T).map(|p| p.as_ref())
-        }
+        // FIXME: alignment
+        unsafe { self.lookup(index).map(|p| p.as_ref()) }
+    }
+
+    #[inline(always)]
+    pub fn get_ptr(&self, index: u32) -> Option<*const T> {
+        // FIXME: alignment
```
Comment (alessandrod) [inline bpf/aya-bpf/src/maps/array.rs:58]: This fixme is unnecessary here

File: bpf/aya-bpf/src/maps/array.rs [hunk 6]
```diff
@@ -47,14 +47,30 @@ impl<T> Array<T> {
...
+        unsafe { self.lookup(index).map(|p| p.as_ptr() as *const T) }
+    }
+
+    #[inline(always)]
+    pub fn get_ptr_mut(&self, index: u32) -> Option<*mut T> {
+        // FIXME: alignment
```
Comment (alessandrod) [inline bpf/aya-bpf/src/maps/array.rs:64]: And here

**Review by alessandrod — APPROVED** [review]

**Pull request merged**